File: BOJ(Java 공부용)/11724.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int[] gox = {0, 1, 0, -1};
    static int[] goy = {1, 0, -1, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        ArrayList<Integer>[] graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) graph[i] = new ArrayList<>();
        boolean[] visited = new boolean[n + 1];
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int first = Integer.parseInt(st.nextToken()), second = Integer.parseInt(st.nextToken());
            graph[first].add(second);
            graph[second].add(first);
        }
        int cnt = 0;
        for (int i = 1; i <= n; i++) {
            if (!visited[i]) {
                dfs(i, graph, visited);
                cnt++;
            }
        }
        System.out.println(cnt);
    }

    private static void dfs(int here, ArrayList<Integer>[] graph, boolean[] visited) {
        visited[here] = true;
        for (int i = 0; i < graph[here].size(); i++) {
            int next = graph[here].get(i);
            if (!visited[next]) dfs(next, graph, visited);
        }
    }
}

File: BOJ(Python 공부용)/11724.py
#!/usr/bin/env python

import sys

sys.setrecursionlimit(500000)
cnt = 0
n, m = map(int, input().split())
graph = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)


def dfs(x):
    visited[x] = True
    for Next in graph[x]:
        if not visited[Next]:
            dfs(Next)


for _ in range(m):
    here, next = map(int, sys.stdin.readline().split())
    graph[here].append(next)
    graph[next].append(here)

for i in range(1, n + 1):
    if not visited[i]:
        cnt += 1
        dfs(i)

print(cnt)
